### DISCUSSION TITLE
How to remove an instance

### DISCUSSION THREAD

**Discussion by hutorny** [eProsima/Fast-DDS#4738] [Q&A]
A writer creates an instance by writing data to its topic. Eventually, that instance may become not relevant and writer needs to remove it, so that existing subscribers get rid of it, and late subscribers would not get an irrelevant instance. Fast DDS API provides method `dispose`, described as 
_This operation requests the middleware to delete the data (the actual deletion is postponed until there is no more use for that data in the whole system)._ 
In my application `dispose` has no effect. I guess, _no more use_ condition is not satisfied. 
Is there a way to forcefully remove an instance?

**Comment by JesusPoderoso** [comment]
Hi @hutorny, thanks for using Fast DDS.
First of all, make sure that your [topic has keys](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/topic/typeSupport/typeSupport.html#data-types-with-a-key). 

> In my application `dispose` has no effect. I guess, _no more use_ condition is not satisfied. 

That may be caused by a Reader that has not performed the `take` call yet.

> ... so that existing subscribers get rid of it, and late subscribers would not get an irrelevant instance.

You should also take into account that the [`dispose()`](https://fast-dds.docs.eprosima.com/en/latest/fastdds/python_api_reference/dds_pim/publisher/datawriter.html#fastdds.DataWriter.dispose) method does not avoid late joiners to receive information from the disposed instance, meanwhile [`unregister_instance()`](https://fast-dds.docs.eprosima.com/en/latest/fastdds/python_api_reference/dds_pim/publisher/datawriter.html#fastdds.DataWriter.unregister_instance) method does.
